scripts/runnable_recipe_ray_vllm_weight_sync.py

```diff
@@ -368,6 +369,447 @@ def run(self):
+                self.prompt_logprobs = torch.tensor(
+                    [
+                        v[tid].logprob if v is not None else 0.0
+                        for v, tid in _zip_strict(
+                            self.prompt_logprobs, self.prompt_token_ids
+                        )
+                    ]
+                )
+            self.prompt_token_ids = torch.tensor(self.prompt_token_ids)
+            self.num_cached_tokens = torch.tensor(self.num_cached_tokens)
+
+    @classmethod
+    def from_request_output(cls, requests):
+        out = lazy_stack(
+            [
+                cls(
+                    request_id=request.request_id,
+                    prompt=request.prompt,
+                    prompt_token_ids=request.prompt_token_ids,
+                    prompt_logprobs=request.prompt_logprobs,
+                    outputs=request.outputs,
+                    finished=request.finished,
+                    metrics=request.metrics,
+                    lora_request=request.lora_request,
+                    encoder_prompt=request.encoder_prompt,
+                    encoder_prompt_token_ids=request.encoder_prompt_token_ids,
+                    num_cached_tokens=request.num_cached_tokens,
+                )
+                for request in requests
+            ]
+        )
+        return out
+
+def _get_output_tokens_and_log_probs(tokens_out, tokenizer, log_prob_key, token_response_key, text_response_key, padding_value):
+    tokens_out = _RequestOutput_tc.from_request_output(tokens_out)
+
+    tokens_response_td = tokens_out.outputs._tensordict.select(
+        "text", "token_ids", "logprobs", strict=False
+    )
+
+    # ================ added padding ===========================================
+    tokens_response_td = tokens_response_td.densify(
+                layout=torch.strided
+    ).to_padded_tensor(padding=padding_value)
+
+    padded_values = tokens_response_td["token_ids"] == padding_value
+    if padded_values.any():
+        lps = tokens_response_td["logprobs"]
+        lps = torch.where(expand_as_right(~padded_values, lps), lps, 1.0)
+        tokens_response_td["logprobs"] = lps
+    # ==========================================================================
+    
+    
+    tokens_response_td.rename_key_("token_ids", token_response_key)
+    tokens_response_td.rename_key_("text", text_response_key)
+    tokens_response_td.rename_key_("logprobs", log_prob_key)
+
+    return tokens_response_td
+
+# =============================================================================================
+
+
+# not decorating with @ray.remote here because num_gpus should vary based on tp_size
+class LLMCollector(SyncDataCollector):
+    """A simplified version of SyncDataCollector for LLM inference."""
+
+    def __init__(
+        self,
+        cfg,
+        llm,
+        # env,
+        policy,
+        queue,
+        worker_id,
+        *,
+        steps_per_batch: int = -1,
```

Comment:
What is a "step" here? A dialog turn?

Reply:
That's my understanding, cc @vmoens to confirm

Reply:
yes!
 Should we rename it?